### DISCUSSION TITLE
SubDyn simulation Node force output problems

### DISCUSSION THREAD

**Discussion by HiHabi** [OpenFAST/openfast#1915] [General]
Dear all, 
I'm currently working on the Openfast simulation and the comparison between the abaqus model. As I'd modify parts of the Jacket Structure in SubDyn and the entire simulation is idea, however when outputting the dynamic force at certain member as well the nodes, it always displayed 0 for every output node I'd set in the SubDyn inputfile, can anyone help with this problem?

the main task of my problem is to print the dynamic force at the Xjoint, where the nodeID=28, 30
the connecting memberID sets are as follow
MemberID/ Node1ID/ NodeID2/ MPropSetID1/ MPropSetID2/ MType
	31	8	28	5	5	1																																										
	32	14	28	5	5	1
	39	9	28	6	6	1																																										
	40	15	28	6	6	1	
	63	10	30	7	7	1																																										
	64	16	30	7	7	1
	71	11	30	7	7	1																																										
	72	17	30	7	7	1	
=====================
-------------------------	MEMBER	OUTPUT	LIST	------------------------------------------																																												
	8	NMOutputs	-	Number	of	members	whose	forces/displacements/velocities/accelerations	will	be	output	(-)	[Must	be	<=	9].																																
MemberID	NOutCnt	NodeCnt	![NOutCnt=how	many	nodes	to	get	output	for	[<	10];	NodeCnt	are	local	ordinal	numbers	from	the	start	of	the	"member,"	and	must	be	>=1	and	<=	NDiv+1]	If	NMOutputs=0	leave	blank	as	well.													
	(-)	(-)	(-)																																													
	31	1	3	!	M1																																											
	32	1	3	!	M2																																											
	39	1	3	!	M3																																											
	40	1	3	!	M4																																											
	63	1	1	!	M5																																											
	64	1	1	!	M6																																											
	71	1	5	!	M7																																											
	72	1	5	!	M8																																											
-------------------------	SDOutList:	The	next	line(s)	contains	a	list	of	output	parameters	that	will	be	output	in	<rootname>.SD.out	or	<rootname>.out.	------																													
"IntfFXss, IntfFYss, IntfFZss, IntfMXss, IntfMYss, IntfMZss"	-	Total	interface	reaction	forces	and	moments	at	the	TP	reference	point	location	in	SS	coordinate	system																																	
"M5N1FKXe, M5N1FKYe, M5N1FKZe, M5N1MKXe, M5N1MKYe, M5N1MKZe"
"M6N1FKXe, M6N1FKYe, M6N1FKZe, M6N1MKXe, M6N1MKYe, M6N1MKZe"	
"M7N5FKXe, M7N5FKYe, M7N5FKZe, M7N5MKXe, M7N5MKYe, M7N5MKZe"
"M8N5FKXe, M8N5FKYe, M8N5FKZe, M8N5MKXe, M8N5MKYe, M8N5MKZe"																									
END	of	output	channels	and	end	of	file.	(the	word	END	must	appear	in	the	first	3	columns	of	this	line)	
=================

The outputs for all M7N5 and M8N5 are all zeros as second figure.

Please offer any suggestion about fixing this problem, many thanks.

Sincerly,
Jack

![image](https://github.com/OpenFAST/openfast/assets/93511835/93fd8116-4f65-4721-8dd9-dac4385a3c15)

the following is the modified SubDyn file, the Openfast simualtion  seem terminated normally yet the output force from the node are always remain 0. 
[SybDyb_backup.txt](https://github.com/OpenFAST/openfast/files/13555945/SybDyb_backup.txt)
![image](https://github.com/OpenFAST/openfast/assets/93511835/7b52a92b-a346-4200-b20e-149a00bcdc68)

**Comment by jjonkman** [comment]
Dear @HiHabi,

I see two problems with your SubDyn input file.  First, you've only selected one output node (i.e., `NOutCnt` = 1) on each output member, so, you can only select N1 in the `SDOutList` (N2-N9 are invalid and will return zeros).  Second, your model uses only two elements per member (`NDiv` = 2), so each member will have 3 nodes.  So, you can't set `NodeCnt` higher than 3 (whereas you have set some of them to 5).

Best regards,

**Comment by HiHabi** [comment]
Dear @jjonkman 

Thanks for your kind response,
I'd see where the mistakes I'd made, thanks for the clarification.

As I'm understood, as taking the MemberID=71 for instance, where the NodeID2 of the member is 30 as I'd like to output the force values, the modification of the parameters are:
NDiv=5 (the assuming subelements I'd like to simulate, does this makes huge difference between 2 or 5?)
NodeCnt=6, refering from the NDiv=5+1 to print the response at NodeID=30 as this is the final node in the member?

please help with the confusion, many thanks

Sincerly, Habi

**Comment by HiHabi** [comment]
Dear @jjonkman 

As I'd  remain the NDiv=2 for testing, the following pic are the output list of the member force being modified. The adjusted SubDyn input file as well the openfast simulation error are as follow:
![image](https://github.com/OpenFAST/openfast/assets/93511835/e8d71605-50f5-4d99-8f1a-58ec76991a97)
[SybDyb_backup.txt](https://github.com/OpenFAST/openfast/files/13565074/SybDyb_backup.txt)

Still taking the MemberID=71, with NodeID1=11/ NodeID2=30, if I'd like to output the dynamic force at NodeID=30, would you recommend to swap the NodeID in MemberID=71 as NodeID1=30/ NodeID2=11 as solution? Or how may I adjust the output list properties with the valus setting for parameters as NOutCnt/ NodeCnt?

Many thanks for your precious time, 

Sincerly, Jack

**Reply by jjonkman** [reply]
Dear @HiHabi,

`NodeCnt` is simply the a list of nodes from 1 to `NDiv`+1.  When `NDiv`=2, the member has two elements with three nodes, so, `NodeCnt` can be set to 1 (start node), 2 (middle node), and/or 3 (end node).  For member 71, the start node is 11 and end node is 30, so, `NodeCnt` = 1 corresponds to `NodeID`=11 and `NodeCnt`=3 corresponds to `NodeID`=30.

Best regards,

**Comment by HiHabi** [comment]
Dear @jjonkman 

It seem my understandings are correct, however is there a must of changing at the parameter NOutCut from 1 to 3?
I'd test on this parameter for NOutCut=1/ 2/ 3 while remaining the NodeCnt=3 for Member7, as the pattern as M7N3FKXe and so, yet the error of Failed to read member output list properties still shows up in NOutCut=2 and 3, only when the NOutCut=1 that Openfast can simulate properly yet this turn back to the output zero problems as the begin.

Please help with this problem, many thanks.

**Reply by jjonkman** [reply]
Dear @HiHabi,

`NOutCnt` is the number of output nodes for the specific output member.  If `NOutCnt` > 1, you must identify that many values of `NodeCnt`.  For example, if you set `NOutCnt`=2 with `NodeCnt`=1,3 for `MemberID`=71, i.e.: 71 2 1,3, then output node 1 will correspond to `NodeID`=11 of member 71 and output node 2 will correspond to `NodeID`=30 of member 71.

Best regards,

**Comment by HiHabi** [comment]
Dear @jjonkman ,

Thanks again for your kind reply and patients.

As I'd follow your instruction about adjusting the memberoutputlist and the output results are still remaining as zero as I'd like to output the response at NodeID=30 at memberID=71 and 72, as M7N3 and M8N3 from the memberoutput list.
![image](https://github.com/OpenFAST/openfast/assets/93511835/5b7bd45f-c8be-4e1a-9641-ca63f6fe045e)

The NodeID=30 is setted as the 3rd layer Xjoint for the jacket substruct for your reference, however I'd believe the response of this place would not be zero. 

however, if the simulation is submitted with M7N1 and M8N1, it seems work properly as the output of the node dynamic force aren't zero, therefore I believe the problem doesn't result from structure modeling problem. 
Would you please suggest any idea about  solving this problem?

Results for M7N1 and M8N1:
![image](https://github.com/OpenFAST/openfast/assets/93511835/66170821-b51e-446a-82dd-79840113de3a)

Results for M7N3 and M8N3:
![image](https://github.com/OpenFAST/openfast/assets/93511835/e276f5ed-f534-4915-a97b-4d92d4f0cea7)

**Reply by jjonkman** [reply]
Dear @HiHabi,

You've defined `NOutCnt` = 2, so, there are only 2 output nodes per member.  Thus, you can select N1 and N2 in the member output names within the `SSOutList`, but N3 is invalid and will return zeros.  N2 refers to the 3rd node along the member per your setting.

Best regards,

**Comment by HiHabi** [comment]
Dear @jjonkman 

Thanks for your kind reply, the output results are available now. 
However I still have some confusion as follow:

For the Xjoint Structure, i.e. NodeID=30, the surroudning MemberID=63/ 64/ 71/ 72
======================================================
MemberID	MJointID1	MJointID2	MPropSetID1	MPropSetID2	MType
	63	10	30	7	7	1																																										
	64	16	30	7	7	1	
	71	11	30	7	7	1																																										
	72	17	30	7	7	1	
========================
As I print output all the Dynamic response of all members mentioned above as the output node format as:
MemberID/ NOutCnt/ NodeCnt
63/ 2/ 1,3
64/ 2/ 1,3
71/ 2/ 1,3
72/ 2/ 1,3

the responses for each member at NodeID=30 seem varies a lots, is this the correct result for the structure response? As my assumptions are that the surround members output at NodeID=30 shall share similar response for convergence.
For example, F5N2FMXe shall share similar results with F6N2FMXE, F7N2FMXe and F8N2FMXe as all these points are referring to the FX dynamic result at NodeID=30.
The Dynamic responses are give as follow pic, as well the excel file, second sheet.

![image](https://github.com/OpenFAST/openfast/assets/93511835/373c5146-076b-43b1-bc4a-030d53ec9918)
[Node30 4member response.xlsx](https://github.com/OpenFAST/openfast/files/13588641/Node30.4member.response.xlsx)

Please help with this problem, many thanks.

**Reply by jjonkman** [reply]
Dear @HiHabi,

I haven't looked closely at your results, but these loads are measured in different members with different local coordinate systems, so, I would not expect the loads to be the same.

Best regards,

**Reply by HiHabi** [reply]
Dear @jjonkman 

The reason that I'd like to measure the node dynamic force is that I'd create another identical model in abaqus, as I'd like to make some work on comfirming the two results having similar result and making some comparison, that is the main object I'd like to work on the output of NodeID=30.

Is there any suggestion you'd suggest to make the verification, or should I try to make the coordinate transformation for each member results above into global coordinate ones' for comparison.

Sincerely, Jack

**Reply by jjonkman** [reply]
Dear @HiHabi,

The coordinate system is only one reason for the difference.  The other reason is that different members will have different loads, even when the intersect at a joint.

Can't you compare the member-level loads directly between SubDyn and ABAQUS for verification?

Best regards,

**Comment by HiHabi** [comment]
Dear @jjonkman 

As I'd closed out the hydrodyn for the aero elastic oneway coupled simulation in Openfast as well in Abaqus, the displacement for observation points seem matching, however with the hydrodyn added, the difference may vary to a large range. Therefore a further quesiton about how to apply the response of hydrodyn into abaqus will be the further questions I'd like to seek for your advices, is there any suggestiong that I may as well applied the hydrodyn results into the abaqus software that can sharing the similar dynamic behaviours as hydrodyn to Openfast main software?

Thanks for your kind reply and patience,
Sincerely, Jack

**Reply by jjonkman** [reply]
Dear @HiHabi,

In OpenFAST, the hydrodynamic loads are dependent on the structural motion (hydroealastic coupling).  So, it would not be appropriate to take the hydrodynamic loads from OpenFAST and apply them directly to ABAQUS.  Rather, a two-way hydro-elastic coupling would be required.

Best regards,
